idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/ui/dialogs/DialogSyntax.kt

```diff
@@ -0,0 +1,106 @@
+package arrow.meta.ide.dsl.ui.dialogs
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.internal.Noop
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.ide.IdeView
+import com.intellij.ide.actions.CreateFileFromTemplateAction
+import com.intellij.ide.actions.CreateFileFromTemplateDialog
+import com.intellij.ide.fileTemplates.FileTemplate
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.actionSystem.LangDataKeys
+import com.intellij.openapi.actionSystem.PlatformDataKeys
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.ProjectRootManager
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import org.jetbrains.kotlin.idea.KotlinFileType
+import org.jetbrains.kotlin.idea.actions.NewKotlinFileAction
+import javax.swing.Icon
+
+/**
+ * Dialogs are isomorphic to a function `(UserInput) -> (AnAction)`.
+ * Therefore Dialogs have the ability to manipulate the editor environment as extensive as `AnAction`.
+ * Consequently, we can treat `Dialogs` as `AnAction` with a display.
+ */
+interface DialogSyntax {
+  /**
+   * ```kotlin:ank:playground
+   * import arrow.meta.Plugin
+   * import arrow.meta.ide.IdeMetaPlugin
+   * import arrow.meta.invoke
+   * import org.jetbrains.kotlin.idea.KotlinFileType
+   *
+   * val IdeMetaPlugin.createf=FilePlugin: Plugin
+   *  get() = "Create File Dialog" {
+   *   meta(
+   *     addFileAction("ExampleAction", "New File", "Creates a new File",
+   *       buildDialog = { project, directory ->
+   *         setTitle("CostumeTitle")
+   *         addKind("File", KotlinFileType.INSTANCE.icon, "Kotlin File")
+   *       }
+   *      )
+   *     )
+   *   }
+   * ```
+   * @param actionId has to be unique
```

Comment:
I haven't been very good about this part and plan on coming back later on in the quote-and-template system, but should we only describe tricky params or all of them eventually?

Reply:
It depends on which of them are important to know. For example with `actionId`. the ide crashes if there are more than 2 Actions with the same Id. Those `actionIds` are used by the`AnActionManager`, which orchestrates all existing Actions and their lifecycle.